### PR TITLE
coremodels: Automatically generate coremodel registries

### LINKED ISSUE #49370: Add a code generator that keeps the coremodel registry up to date
The coremodel system relies on a registry (`pkg/framework/coremodel/staticregistry`) to enumerate the set of, and provide access to, individual coremodels. Providing such a registry is necessary for cases where some code wants to act on the set of coremodels, rather than any particular one.

Having to update this registry by hand whenever creating a new coremodel is toilsome. It also introduces the possibility of errors where there really need not be any. Consequently, we should rely on codegen to keep the static registry up to date. Might be useful for other registry patterns elsewhere in the codebase, as well.

Note - we don't have any cases currently merged that operate on the set of coremodels. But there are some known cases - for example, i'm about to pop out an as-code library that will depend on them, even though it'll be in a separate repo. #44840 (which is paused) introduced a new HTTP server that needed to act on a set. And one next logical step from #48252 is to make validation by coremodel a generic part of the web framework, which would mean relying on the set.

### PR DESCRIPTION
**Pull request by sdboyer**
**What this PR does / why we need it**:

This one is pretty slick, IMO :)

This extends the code generation introduced in #49193 to _also_ generate the registry that populates and mediates access to coremodels. This is pretty nifty, because it means we've pretty much reached full encapsulation of the coremodel - add a new dir in `pkg/coremodel/*` with a `lineage.cue` file, and the registry gets updated automatically, as well as all the generated code just popping out.

Crucially, with this automated generation, it's now much more reasonable to generate a static registry - that is, one that's amenable to static analysis. We discussed this a bit over in #48252, which initially relied on a string lookup (passing `"dashboard"` as a param) to get the coremodel for validating incoming JSON in the HTTP handler. This PR solves that problem.

I do think it's reasonable to extend this kind of registry generation approach to other contexts.

cc @joeblubaugh 

**Which issue(s) this PR fixes**:

Fixes #49370

**Special notes for your reviewer**:

There's a moderate amount of noise in this PR:

* I changed the standard comment header on verified files. Zoinks.
* There is zero functionality change under `pkg/api` - all that's happening is a) replacing the old string-oriented registry with a call to the static one, and updating hacks in the tests accordingly

### CODE REVIEW

**Comment by sdboyer** [comment]
imma do an example commit tomorrow that shows what all pops out when we add a datasource coremodel

File: pkg/api/dashboard.go [hunk 1]
```diff
@@ -311,22 +311,22 @@ func (hs *HTTPServer) PostDashboard(c *models.ReqContext) response.Response {
 	}
 
 	if hs.Features.IsEnabled(featuremgmt.FlagValidateDashboardsOnSave) {
+		cm := hs.CoremodelStaticRegistry.Dashboard()
+
 		// Ideally, coremodel validation calls would be integrated into the web
 		// framework. But this does the job for now.
-		if cm, has := hs.CoremodelRegistry.Get("dashboard"); has {
```
Comment (sdboyer) [inline pkg/api/dashboard.go:316]: Yay, we no longer rely on string lookups - static analysis ftw!

File: pkg/framework/coremodel/registry/registry_gen.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	"github.com/grafana/grafana/pkg/cuectx"
+	"github.com/grafana/grafana/pkg/framework/coremodel"
+	"github.com/grafana/thema"
+)
+
+// CoremodelSet contains all of the wire-style providers related to coremodels.
+var CoremodelSet = wire.NewSet(
+	ProvideStatic,
+	ProvideGeneric,
+)
```
Comment (sdboyer) [inline pkg/framework/coremodel/registry/registry_gen.go:23]: IMO chunking out logical subsets of wire providers into distinct sets like this - then codegenning them! - is a pretty nifty way to cut down on the massive, monotonically increasing pile of stuff in `pkg/server/wire.go`

File: pkg/framework/coremodel/gen.go [hunk 3]
```diff
@@ -77,6 +77,13 @@ func main() {
 		wd.Merge(wdt)
 	}
 
+	wdr, err := gcgen.GenerateCoremodelRegistry(filepath.Join(groot, "pkg", "framework", "coremodel", "registry", "registry_gen.go"), lins)
```
Comment (malcolmholmes) [inline pkg/framework/coremodel/gen.go:80]: you could do with using more descriptive naming here. What is `wd`? `wdr`? `wtd`?

Comment (malcolmholmes) [reply pkg/framework/coremodel/gen.go:80]: I get that this is generated code, but it still used and still needs to be understandable.

Comment (sdboyer) [reply pkg/framework/coremodel/gen.go:80]: This is the code generator, not generated code, but either way, fair enough.

`wd` is the overall `WriteDiffer`. i've left that one as-is. The others, `wdg`, wdt`, `wdr`, are `WriteDiffer`s produced by sub-generators, for Go, Typescript, and Registry, respectively. They all get merged into the main `wd`, which effectively batches their writes together.

i've renamed those latter three to be more suggestive.

**Review by malcolmholmes — APPROVED** [review]

**Pull request merged**